File: creation-validation.mjs
const MAX_BYTES=1050000, MAX_URL=1400023;
export function decodePreview(value){
  if(typeof value!=='string'||value.length>MAX_URL||!value.startsWith('data:image/jpeg;base64,'))throw new Error('Invalid preview');
  const encoded=value.slice(23);
  if(!encoded || !/^[A-Za-z0-9+/]+={0,2}$/.test(encoded)||encoded.length%4!==0)throw new Error('Invalid base64');
  const raw=Buffer.from(encoded,'base64');
  if(raw.length<4||raw.length>MAX_BYTES||raw[0]!==255||raw[1]!==216||raw[2]!==255||raw.at(-2)!==255||raw.at(-1)!==217)throw new Error('Invalid JPEG');
  return raw;
}
export function validateCreations(creations){
  if(!Array.isArray(creations)||creations.length>2)throw new Error('Too many creations');
  const kinds=new Set(),summaries=[],attachments=[];
  for(const c of creations){
    if(!c||!['shirt','tube'].includes(c.kind)||kinds.has(c.kind)||typeof c.summary!=='string'||!c.summary.trim()||c.summary.length>3500)throw new Error('Invalid creation');
    kinds.add(c.kind);summaries.push((c.kind==='shirt'?'T-SHIRT — TEXTILE NON FOURNI, FOURNI PAR LE CLIENT.':'TUBE PERSONNALISÉ.')+'\n'+c.summary.trim());
    attachments.push({name:`badonline-${c.kind}.jpg`,bytes:decodePreview(c.preview)});
    if(c.photo!==undefined){if(c.kind!=='tube')throw new Error('Unexpected photo');attachments.push({name:'badonline-photo-tube.jpg',bytes:decodePreview(c.photo)});}
  }
  return {summary:summaries.join('\n\n'),attachments};
}

File: creation_validation.py
"""Validate small JPEG design previews without trusting browser filenames or recipients."""
import base64
import binascii

MAX_JPEG_BYTES = 1050000
MAX_DATA_URL = 1400023


def decode_preview(value):
    if not isinstance(value, str) or len(value) > MAX_DATA_URL or not value.startswith('data:image/jpeg;base64,'):
        raise ValueError('Aperçu invalide ou trop volumineux.')
    try:
        raw = base64.b64decode(value.split(',', 1)[1], validate=True)
    except (binascii.Error, ValueError):
        raise ValueError('Aperçu invalide.') from None
    if not 4 <= len(raw) <= MAX_JPEG_BYTES or not raw.startswith(b'\xff\xd8\xff') or not raw.endswith(b'\xff\xd9'):
        raise ValueError('Format d’aperçu invalide.')
    return raw


def validate_creations(creations):
    if not isinstance(creations, list) or len(creations) > 2:
        raise ValueError('Vous pouvez joindre un t-shirt et un tube par demande.')
    summaries, attachments, kinds = [], [], set()
    for creation in creations:
        if not isinstance(creation, dict) or creation.get('kind') not in ('shirt', 'tube'):
            raise ValueError('Création invalide.')
        kind = creation['kind']
        if kind in kinds:
            raise ValueError('Création en double.')
        kinds.add(kind)
        summary = creation.get('summary')
        if not isinstance(summary, str) or not 1 <= len(summary.strip()) <= 3500:
            raise ValueError('Description de création invalide.')
        prefix = 'T-SHIRT — TEXTILE NON FOURNI, FOURNI PAR LE CLIENT.' if kind == 'shirt' else 'TUBE PERSONNALISÉ.'
        summaries.append(prefix + '\n' + summary.strip())
        attachments.append((f'badonline-{kind}.jpg', decode_preview(creation.get('preview'))))
        if creation.get('photo') is not None:
            if kind != 'tube':
                raise ValueError('Photo inattendue.')
            attachments.append(('badonline-photo-tube.jpg', decode_preview(creation['photo'])))
    return '\n\n'.join(summaries), attachments
